ENH: Environment class major refactor may 2024

Gui-FernandesBR opened the pull request:
## Pull request type

- [x] Code maintenance (refactoring, formatting, tests)

## Checklist

- [x] Tests for the changes have been added (if needed)
- [x] Docs have been reviewed and added / updated
- [x] Lint (`black rocketpy/ tests/`) has passed locally 
- [x] All tests (`pytest tests -m slow --runslow`) have passed locally
- [x] `CHANGELOG.md` has been updated (if relevant)

## Current behavior

Reading and maintaining the Environment class is not an easy task.

## New behavior

First of all, I'm reeeeeeeeally sorry for the size of this PR.  These are the overall highlights of my changes:

### Enhancements (ENH)
- Created rocketpy.environment.tools module.
- Created the environment.fetchers module.
- Created a `WeatherModelMapping` class.
- Added modulo operator to Function class.
- Added auxiliary private setters to the Environment class.
- Added a few input validation functions to the Environment class.

###  Documentation (DOC)
- Reviwed the entire class docstrings!! Many errors were fixed.

###  Bug Fixes (BUG)
- The CMC Ensemble model is no longer available, so I removed it and updated the tests. In the future I can re-implement it with the new API endpoint.

### Maintenance (MNT)
- Refactored almost 99% of the Environment class.
- Fixed pylint and flake8 errors in the code.
- Code is cleaner, lighter and more readable! The environment.py file now has less than 3l code lines. 

### Testing (TST)
- Updated all the Environment tests.
- Re-evaluated slow tests and optimized them.
- **I reduced the execution time and increased the coverage of the tests.** 


## Breaking change

- [x] No (at least not hard breaking changes).

## Additional information

Now it's official... We have remodeled all the original RocketPy classes.

### Review:

To better review this PR I recommend:

- Run the slow tests locally (if it fails, please comment it here so we can reproduce it.)
- Run the `getting_started` notebook 
- Run the Environment class docstrings and usage notebook.
- Read the codecoverage report and search for potential gaps that are not being covered by the tests
- You can ignore all the modifications in the `tests` module, since it is easy to fix it later.
- Remember that we have pylint and flake8 already passing in the CI
- Try to playaround... Import the environment class and create different objects simulating the user behavior. 
- If you have to prioritize which files to read, I recommend you to read the new files that are being added.

MateusStano commented on rocketpy/environment/environment.py:
```diff
@@ -3083,7 +2393,7 @@ def calculate_density_profile(self):
         >>> float(env.density(1000))
-        1.1116193933422585
+        1.1115112430077818
         """
```
Why did this change?

Gui-FernandesBR replied:
I honestly don`t know. The difference was so small that I ignored it for a moment.
However, I'm curious now.

Maybe it was related to the discretization of barometric pressure somehow.

MateusStano commented on rocketpy/environment/environment.py:
```diff
@@ -3029,34 +2328,45 @@ def pressure_function(h):
+
+        return np.column_stack([altitudes, pressures])
+
+    @funcify_method("Pressure (Pa)", "Height Above Sea Level (m)")
+    def barometric_height_ISA(self):
+        """Returns the inverse function of the pressure_ISA function."""
+        return self.pressure_ISA.inverse_function()
+
```
This is not setting discrete like in the previous code. Why?

Gui-FernandesBR replied:
The `barometric_height_ISA` is already discretized, therefore the set_discretize() call was not necessary in here.

MateusStano commented on rocketpy/environment/environment.py:
```diff
@@ -372,44 +446,177 @@ def __init__(
+            interpolation="linear",
         )
 
+    def __reset_barometric_height_function(self):
+        # NOTE: this assumes self.pressure and max_expected_height are already set.
+        self.barometric_height = self.pressure.inverse_function()
+        if callable(self.barometric_height.source):
+            # discretize to speed up flight simulation
+            self.barometric_height.set_discrete(
+                0,
+                self.max_expected_height,
+                100,
+                extrapolation="constant",
+                mutate_self=True,
+            )
+        self.barometric_height.set_inputs("Pressure (Pa)")
+        self.barometric_height.set_outputs("Height Above Sea Level (m)")
```
`__reset_barometric_height` uses set_discrete, but `__set_barometric_height` does not?

With the way you used it, this seems to be correct, but just checking in case this is not intended

Gui-FernandesBR replied:
Ok, I see it now. I will make sure the `__set_barometric_height` also discretizes the Function if needed.

Lucas-Prates commented on rocketpy/environment/fetchers.py:
```diff
@@ -0,0 +1,433 @@
+        The latitude of the location.
+    lon : float
+        The longitude of the location.
+    model : str
+        The atmospheric model to use. Options are: ecmwf, GFS, ICON or ICONEU.
+
+    Returns
+    -------
+    dict
+        A dictionary containing the atmospheric data retrieved from the API.
+
+    Raises
+    ------
+    ValueError
+        If an invalid response is received from the API.
+    """
+    model = model.lower()
+    if model[-1] == "u":  # case iconEu
+        model = "".join([model[:4], model[4].upper(), model[5:]])
+
+    url = (
+        f"https://node.windy.com/forecast/meteogram/{model}/{lat}/{lon}/"
+        "?step=undefined"
+    )
+
+    try:
+        response = requests.get(url).json()
+        if "data" not in response.keys():
+            raise ValueError(
+                f"Could not get a valid response for {model} from Windy. "
+                "Check if the coordinates are set inside the model's domain."
+            )
+    except requests.exceptions.RequestException as e:
+        if model == "iconEu":
+            raise ValueError(
+                "Could not get a valid response for Icon-EU from Windy. "
+                "Check if the coordinates are set inside Europe."
+            ) from e
+
+    return response
```
When the user incorrectly specifies the `model`, i.e. it is not a valid model name , the fetcher runs for a whole 1 minute and then returns a `UnboundLocalError` exception. Maybe we should check beforehand if the model name is correct and raise a more informative exception.

Below is a code to reproduce the error. Using "ECMWF" for `file` runs correctly, but with the incorrect model name it gives the incorrect behavior described above. Note that the `file` argument for `set_atmospheric_model` is eventually passed as the `model` argument for `fetch_atmospheric_data_from_windy`.

```
EnvWindy = Environment(
    date=(2019, 2, 10, 18),
    latitude=-21.960641,
    longitude=-47.482122,
)
EnvWindy.set_atmospheric_model(
    type="Windy",
    file="oh_no_wrong_model_name",
    #file="ECMWF",
)
```

Gui-FernandesBR replied:
Hey, this is a nice observation.
I added a validation to the Environemnt class, before calling this fetcher function.

I think it is not a good idea to raise an exception inside the fetcher since it would still run indefinitelly until the max_attemp count reaches the limit.

Lucas-Prates commented on rocketpy/environment/fetchers.py:
```diff
@@ -0,0 +1,433 @@
+    attempt_count = 0
+    dataset = None
+
+    while attempt_count < max_attempts:
```
Isn't this functionality provided by the `exponential_backoff` decorator? This repeats in other fetchers, some of which use the mentioned decorator.

Gui-FernandesBR replied:
Your question id good.

I don`t really like this behavior...
But what the code is doing is to actually use the attempt count to find the latest data available on the server.

We should REALLY do this retrieval in another way... But I would leave this for another PR. 

I will leave a TODO in the code so we dont forget.

Lucas-Prates commented:
I have given only a very small partial review, I will complete it in time. Here are some points other than the few comments made in code:

1) There was a single assertion error wen running pytest-slow, see below;
![image](https://github.com/RocketPy-Team/RocketPy/assets/57069366/929f956c-0a5e-4507-9f17-f354bce88c6f)

2) The getting started notebook ran smoothly;

3) The environment analysis usage notebook has a CMC Ensemble section that does not run. From what I understood from the documentation of `set_atmospheric_model`, CMC Ensembles are currently unavailable;

4) I mostly "played around" with stuff, so I haven't read thoroughly the code yet.

Gui-FernandesBR commented on rocketpy/environment/__init__.py:
```diff
@@ -1,2 +1,9 @@
+
 from .environment import Environment
 from .environment_analysis import EnvironmentAnalysis
+
+__all__ = ["Environment", "EnvironmentAnalysis"]
```
O understand it is not mandatory. But I was trying to enforce that these are the only 2 public classes within the `rocketpy.envrionment` module.

See the module docstring: 
![alt text](image.png)

If you still think it is a problem to define the __all__ variable here, please let me know.

Gui-FernandesBR commented on rocketpy/environment/tools.py:
```diff
@@ -0,0 +1,684 @@
+
+    Returns
+    -------
+    numpy.ma.MaskedArray
+        A cleaned array with rows containing masked values removed.
+
+    Raises
+    ------
+    UserWarning
+        If any values were missing and rows were removed.
+    """
+    data_array = np.ma.column_stack(list(args))
```
I dont even know why we document raises at all!!!!!!!!

gonna delete it soon. Thanks for pointing this out.

Gui-FernandesBR commented on rocketpy/environment/tools.py:
```diff
@@ -0,0 +1,684 @@
+    lon_list : list of float
+        The list of longitudes.
+
+    Returns
+    -------
+    tuple
+        A tuple containing the adjusted longitude and its index in the list.
+
+    Raises
+    ------
+    ValueError
+        If the longitude is not within the range covered by the list.
+    """
+    # Determine if file uses -180 to 180 or 0 to 360
+    if lon_list[0] < 0 or lon_list[-1] < 0:
+        # Convert input to -180 - 180
+        lon = longitude if longitude < 180 else -180 + longitude % 180
+    else:
+        # Convert input to 0 - 360
+        lon = longitude % 360
+    # Check if reversed or sorted
+    if lon_list[0] < lon_list[-1]:
+        # Deal with sorted lon_list
+        lon_index = bisect.bisect(lon_list, lon)
+    else:
+        # Deal with reversed lon_list
+        lon_list.reverse()
+        lon_index = len(lon_list) - bisect.bisect_left(lon_list, lon)
+        lon_list.reverse()
+    # Take care of longitude value equal to maximum longitude in the grid
+    if lon_index == len(lon_list) and lon_list[lon_index - 1] == lon:
+        lon_index = lon_index - 1
+    # Check if longitude value is inside the grid
+    if lon_index == 0 or lon_index == len(lon_list):
+        raise ValueError(
+            f"Longitude {lon} not inside region covered by file, which is "
+            f"from {lon_list[0]} to {lon_list[-1]}."
+        )
+
+    return lon, lon_index
```
Given that I did not add these lines, only moved them, I will not try so hard to create a new test to cover these lines.

Gui-FernandesBR commented on rocketpy/environment/tools.py:
```diff
@@ -0,0 +1,684 @@
+
+def find_latitude_index(latitude, lat_list):
+    """Finds the index of the given latitude in a list of latitudes.
+
+    Parameters
+    ----------
+    latitude : float
+        The latitude to find in the list.
+    lat_list : list of float
+        The list of latitudes.
+
+    Returns
+    -------
+    tuple
+        A tuple containing the latitude and its index in the list.
+
+    Raises
+    ------
+    ValueError
+        If the latitude is not within the range covered by the list.
+    """
+    # Check if reversed or sorted
+    if lat_list[0] < lat_list[-1]:
+        # Deal with sorted lat_list
+        lat_index = bisect.bisect(lat_list, latitude)
+    else:
+        # Deal with reversed lat_list
+        lat_list.reverse()
+        lat_index = len(lat_list) - bisect.bisect_left(lat_list, latitude)
+        lat_list.reverse()
+    # Take care of latitude value equal to maximum longitude in the grid
+    if lat_index == len(lat_list) and lat_list[lat_index - 1] == latitude:
+        lat_index = lat_index - 1
+    # Check if latitude value is inside the grid
+    if lat_index == 0 or lat_index == len(lat_list):
+        raise ValueError(
+            f"Latitude {latitude} not inside region covered by file, "
+            f"which is from {lat_list[0]} to {lat_list[-1]}."
+        )
+    return latitude, lat_index
```
Once again. These lines are not new.
It would be nice to have different examples for a better coverage.
However, it is not a problem introduced by this PR

Gui-FernandesBR commented on rocketpy/environment/tools.py:
```diff
@@ -0,0 +1,684 @@
+    return data_array
+
+
+def apply_bilinear_interpolation(x, y, x1, x2, y1, y2, data):
+    """Applies bilinear interpolation to the given data points.
+
```
You're right. I can do much better than this.

Gui-FernandesBR replied:
I now see that I did those functions to "spare some lines".. What a dumb move...

I rolled back and made it better now. I think it is nice that we are using the bilinear interoplation in many places of our repo now!

Gui-FernandesBR commented on rocketpy/environment/tools.py:
```diff
@@ -0,0 +1,684 @@
+def utm_to_geodesic(  # pylint: disable=too-many-locals,too-many-statements
+    x, y, utm_zone, hemis, semi_major_axis=6378137.0, flattening=1 / 298.257223563
+):
+    # NOTE: already documented in the Environment class.
+    # TODO: deprecated the static method from the environment class, use only this one.
```
I agree it is better to do it ASAP.

However, I am avoiding adding more and more commits to this PR.

I prefer to do it later in another PR (or someone else could also do it.)

Gui-FernandesBR commented on rocketpy/environment/fetchers.py:
```diff
@@ -0,0 +1,433 @@
+    )
+    try:
+        response = requests.get(request_url)
+    except requests.exceptions.RequestException as e:
+        raise RuntimeError("Unable to reach Open-Elevation API servers.") from e
```
Thanks for pointing this out. I think I will make the except more broad so it captures more scenarios

Gui-FernandesBR replied:
I'm really sorry to say that I tried it more than 100 times and could not get to an error with Open-Elevation API.
I can't reproduce what you said about the JSONDecodeError.

Something that would help **a lot** is a simple screenshot with the error.

All in all I think Open-Elevation is bare used across our project. I think we should keep it the way it is and improve the retry strategy in the future if an user reports troube.

You can think that before this refactor the open-elevation fetching process did not have any retry policy at all. Now we are at least improving this behavior...

The code I used to test:
```python
from rocketpy import Environment
import random

def main(num_iter=1):
    for i in range(num_iter):
        
        lat = random.uniform(-90, 90)
        lon = random.uniform(-180, 180)
        env = Environment(date=(2019, 2, 10, 18), latitude=lat, longitude=lon)
        env.set_elevation("Open-Elevation")

        env.prints.launch_site_details()


if __name__ == "__main__":
    main(100)

```

Gui-FernandesBR replied:
[e67d988](https://github.com/RocketPy-Team/RocketPy/pull/605/commits/e67d988442f5b8bdd972e0b1713b9c8c6ee85d25)

Gui-FernandesBR commented:
> I have given only a very small partial review, I will complete it in time. Here are some points other than the few comments made in code:
> 
> 1. There was a single assertion error wen running pytest-slow, see below;
>    ![image](https://private-user-images.githubusercontent.com/57069366/347125025-929f956c-0a5e-4507-9f17-f354bce88c6f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjA1ODU2MjYsIm5iZiI6MTcyMDU4NTMyNiwicGF0aCI6Ii81NzA2OTM2Ni8zNDcxMjUwMjUtOTI5Zjk1NmMtMGE1ZS00NTA3LTlmMTctZjM1NGJjZTg4YzZmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MTAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzEwVDA0MjIwNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTA3NGNmOTM3MWI5YzgxNmI1MTU1MzI5NmM4MzU0MTM2NjRiY2ZiMDA2YjY0YWM2Y2NkYzJiYmVjNDE1Y2M0OTkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.4fszhHEPXSno4fyE_lmmh_OQYFyt3pAnd80RQVS25k8)

I can reproduce this error in my machine.
This is odd and it is not related to this PR. I guess it was introduced in the #628, given the commit history.
I think I will fix the test in this PR and we can move on.
We didn`t change anything in the EnvironmentAnalysis class, so it is for sure a problem with the test itself.

> 2. The getting started notebook ran smoothly;
> 3. The environment analysis usage notebook has a CMC Ensemble section that does not run. From what I understood from the documentation of `set_atmospheric_model`, since CMC Ensembles are currently unavailable, should this section be removed?

I will make another PR to update environment documentation. Work in progress.

> 4. I mostly "played around" with stuff, so I haven't read thoroughly the code yet.


Thanks for your review!! Your points were really helpful.

Gui-FernandesBR commented:
@MateusStano @Lucas-Prates I have addressed all of your comments.

Please revisit the resolved conversations and approve the PR if you don't have additional comments.
